### PR TITLE
Move timeouts to client state

### PR DESCRIPTION
**Pull request by psarna**
This series is extracted from #7913 as it may prove useful to other series as well, and #7913 might take a while until its merged, given that it also depends on other unmerged pull requests.

The idea of this series is to move timeouts to the client state, which will allow changing them independently for each session - e.g. by setting per-service-level timeouts and initializing the values from attached service levels (see #7867).

### CODE REVIEW

**Comment by psarna** [comment]
@kbr- I also changed one of your recent commits during a rebase, and now an internally generated CDC request uses per-query `USING TIMEOUT 60s` instead of passing a custom timeout config - plz check if it looks alright.

File: service/client_state.hh [hunk 1]
```diff
@@ -136,6 +144,10 @@ private:
     // Only populated for external client state.
     auth::service* _auth_service{nullptr};
 
+    // For restoring default values in the timeout config
```
Comment (nyh) [inline service/client_state.hh:147]: I don't understand this comment. Why do we need both these timeout configs, and why is the "default" per connection as well?

File: alternator/executor.cc [hunk 2]
```diff
@@ -3214,7 +3214,7 @@ static future<executor::request_return_type> do_query(service::storage_proxy& pr
     auto query_state_ptr = std::make_unique<service::query_state>(client_state, trace_state, std::move(permit));
 
     command->slice.options.set<query::partition_slice::option::allow_short_read>();
-    auto query_options = std::make_unique<cql3::query_options>(cl, infinite_timeout_config, std::vector<cql3::raw_value>{});
```
Comment (nyh) [inline alternator/executor.cc:3217]: @psarna do you remember why we had here an infinite timeout???

File: db/system_distributed_keyspace.cc [hunk 3]
```diff
@@ -526,11 +529,10 @@ future<std::vector<db_clock::time_point>>
 system_distributed_keyspace::get_cdc_desc_v1_timestamps(context ctx) {
     std::vector<db_clock::time_point> res;
     co_await _qp.query_internal(
-            format("SELECT time FROM {}.{}", NAME, CDC_DESC_V1),
-            quorum_if_many(ctx.num_token_owners),
             // This is a long and expensive scan (mostly due to #8061).
             // Give it a bit more time than usual.
-            get_timeout_config(std::chrono::seconds(60)),
```
Comment (kbr-) [inline db/system_distributed_keyspace.cc:533]: For consistency - since everyone uses an especially-crafted `query_state` - we could make a function, `get_query_state(duration)` and use `get_query_state(seconds(60))` here, and define `internal_distributed_query_state` as `get_query_state(seconds(10))`.

But that's just a personal preference, so LGTM.

**Review by nyh — COMMENTED** [review]
I have some small comments, but I'm afraid that I didn't understand the *goal* of this patch. The idea that the timeout is per-connection is a CQL concept (in fact, a CQL concept we're trying to invent...). Other front-ends might require a different timeout per query. So why are you removing the per-query timeout feature? Also, the entire notion of a "connection" and that authentication is done only once per connection is CQL-specific (it's not true in DynamoDB API), and so is "timeout_config" and its pretty specific selection of timeout parameters.
And if I mention Alternator, I didn't understand what this change does to Alternator. Alternator currently has a fixed default_timeout() in code, of 10s (except one place I saw suddenly that used an infinite timeout, what was the intention there??). Will the still be the case after this patch? Will Alternator be simply ignoring this new client_state thing, in which case, maybe it's a wrong abstraction?

**Comment by psarna** [comment]
@nyh Oh no, I'm very far from removing the per-query-timeout feature!!!!11 I only just introduced it :) Historically, we only had a notion of per-coordinator timeout, which meant that timeouts were configured in the yaml file. Later, per-query timeouts were introduced and they're very useful, but we also want to be able to set timeouts with coarser granularity, without having to changing existing queries. One possible use case for that is having multiple Spark jobs - it might be hard to modify the internals of these jobs to provide per-query timeouts, but it's easy to create multiple roles, attach service levels to them and then assign each job its corresponding role.

This pull request is just a first step towards this goal. Right now each client state keeps a reference to the global timeout configuration, and what I'm doing is localizing it, so that each connection has a copy of the timeout config, and is allowed to modify it as it wishes. Actual mechanisms for modifying these timeout configurations will come in later series.

File: service/client_state.hh (hunk 1 above)
Comment (psarna) [reply service/client_state.hh:147]: Let's assume for a while that we already have a per-service-level timeout mechanism from #7913. Now, a user can do:
```cql
ALTER SERVICE LEVEL sl SET read_timeout = 5s;
```
and later:
```cql
ALTER SERVICE LEVEL sl SET read_timeout = null;
```
What the second statement does is removing the custom timeout value, and in this case the session is expected to fall back to its default timeout configuration, which is the one defined in `scylla.yaml`. In order to avoid having to consult database config, a local copy of default values is kept.

Comment (psarna) [reply service/client_state.hh:147]: I think this change is not strictly needed for this pull request though, so I can remove it for now. It's only needed in the broader context of #7913.

File: alternator/executor.cc (hunk 2 above)
Comment (psarna) [reply alternator/executor.cc:3217]: Not really, and it doesn't sound right. It's likely it's simply a bug.

**Comment by avikivity** [comment]
@nyh all protocols have (or should have) the concept of a session, a communication between a client and the cluster that encompasses a large number of queries. Instead of burdening each query with lots of little parameters, we should push those parameters to the session level where it makes sense.

**Comment by nyh** [comment]
@psarna I probably misread your patch, I'll have to read it again... I got the feeling that you are actively removing from some places in the code the ability to have a different timeout for each query, but I guess I misunderstood the code. 

@avikivity you are probably right about what protocols "should" do, but people who listened too much when their CS professor told them about the OSI network model sometimes like to layer their protocol in a way that makes them lovely to describe as distinct layers, but inefficient. I think this is what happened to DynamoDB's protocol. It's designed to be a protocol **over** HTTP (or HTTPS). This means that each request is completely self-contained (all the information on who you are and what you want) - there is no notion of a "connection" or a "connection state". Sure, in the TCP layer there is a connection, but in the HTTP layer, you're not supposed to know about it: When you send two HTTP requests, you are not supposed to care if they are sent over the same connection or two different connections. Or even if the HTTP request when through HTTP proxies, and so on. So in the DynamoDB API each request is self-contained: It is signed individually. If an HTTP proxy is used (for example), it might very well forward two different requests from two different users over the same TCP connection to the destination endpoint. 

So to summarize, I do think that having an option for a per-connection timeout is good, but we need to be careful not to assume that all front-ends will actually want to use this option. I don't think it's bad that we add an option that Alternator won't use. I'm just worried maybe it's a sign that if Alternator also needs to use this object with options that it doesn't want to use, than maybe the abstraction is wrong. But I'm not sure what is a better approach. It is very likely I just don't understand the full picture, and Piotr who does probably did it well.

**Comment by gleb-cloudius** [comment]
On Tue, Feb 23, 2021 at 03:16:38PM -0800, nyh wrote:
> @avikivity you are probably right about what protocols "should" do, but people who listened too much when their CS professor told them about the OSI network model sometimes like to layer their protocol in a way that makes them lovely to describe as distinct layers, but inefficient. I think this is what happened to DynamoDB's protocol. It's designed to be a protocol **over** HTTP (or HTTPS). This means that each request is completely self-contained (all the information on who you are and what you want) - there is no notion of a "connection" or a "connection state". Sure, in the TCP layer there is a connection, but in the HTTP layer, you're not supposed to know about it: When you send two HTTP requests, you are not supposed to care if they are sent over the same connection or two different connections. Or even if the HTTP request when through HTTP proxies, and so on. So in the DynamoDB API each request is self-contained: It is signed individually. If an HTTP proxy is used (for example), it might very well forward two different requests from two different users over the same TCP connection to the destination endpoint. 
> 
It is also much easier to load balance requests when each one has the
whole state. Paging state in CQL request/reply is an example of that.

--
			Gleb.

**Comment by avikivity** [comment]
@nyh we can give alternator a session concept through authentication. Each user is associated with a session type, and each request is associated with a user. Ergo, each request is associated with a session type.

It's not any different from CQL, where a session is spread over many connections, and where we also decided to use the user name to select the session type.

**Comment by nyh** [comment]
> @nyh we can give alternator a session concept through authentication. Each user is associated with a session type, and each request is associated with a user. Ergo, each request is associated with a session type.
> 
> It's not any different from CQL, where a session is spread over many connections, and where we also decided to use the user name to select the session type.

True. I'm just not sure the "client_state" object carries this notion of "session type" and not something else. But maybe it does and I don't understand completely what this "client_state" object is supposed to be.

In any case, there is really no big problem in doing something now, which fits our current needs, and if we want to refactor it yet again at some future time when new needs arise, we can do it then.

File: service/client_state.hh [hunk 4]
```diff
@@ -88,10 +89,17 @@ public:
     };
 private:
     client_state(const client_state* cs, seastar::sharded<auth::service>* auth_service)
-            : _keyspace(cs->_keyspace),  _user(cs->_user), _auth_state(cs->_auth_state),
-              _is_internal(cs->_is_internal), _is_thrift(cs->_is_thrift), _remote_address(cs->_remote_address),
-              _auth_service(auth_service ? &auth_service->local() : nullptr),
-              _enabled_protocol_extensions(cs->_enabled_protocol_extensions) {}
+            : _keyspace(cs->_keyspace)
+            ,  _user(cs->_user)
```
Comment (avikivity) [inline service/client_state.hh:93]: An extra space snuck in before _user, increasing latency.

File: service/client_state.hh (hunk 1 above)
Comment (avikivity) [reply service/client_state.hh:147]: Could we not merge _default_timeout_config into _timeout_config, so we only need one? So the lookup is fast.

File: cql3/statements/select_statement.cc [hunk 5]
```diff
@@ -694,7 +694,7 @@ select_statement::execute(service::storage_proxy& proxy,
     // is specified we need to get "limit" rows from each partition since there
     // is no way to tell which of these rows belong to the query result before
     // doing post-query ordering.
-    auto timeout = db::timeout_clock::now() + get_timeout(options);
+    auto timeout = db::timeout_clock::now() + get_timeout(state.get_client_state(), options);
```
Comment (avikivity) [inline cql3/statements/select_statement.cc:697]: Another way to go would be to set query_options::_timeout_config to point to the client state's _timeout_config. Not saying it's better, just pointing out the alternative.

**Comment by avikivity** [comment]
Looks good, but there is discussion so letting people voice any objections.

File: service/client_state.hh (hunk 1 above)
Comment (psarna) [reply service/client_state.hh:147]: Which lookup? We only use `_timeout_config` for getting proper timeout values. `_default_timeout_config` is here in order to know what to put in `_timeout_config` in case somebody removes a custom timeout value from their config (e.g. by updating their service level parameters). When they do that, this value is set to its corresponding value from `_default_timeout_config`.

File: service/client_state.hh (hunk 4 above)
Comment (psarna) [reply service/client_state.hh]: done, the latency regression is fixed

**Comment by psarna** [comment]
(ping)

**Comment by avikivity** [comment]
I guess the discussion died down.
